src/tribler-core/tribler_core/modules/tunnel/community/triblertunnel_community.py

```diff
@@ -340,29 +304,61 @@ def update_torrent(self, peers, download):
             if self.find_circuits():
                 self.readd_bittorrent_peers()
 
-    def do_payout(self, peer, circuit_id, amount, base_amount):
+    def do_payout(self, peer: Peer, circuit_id: int, amount: int, base_amount: int) -> None:
         """
         Perform a payout to a specific peer.
         :param peer: The peer to perform the payout to, usually the next node in the circuit.
         :param circuit_id: The circuit id of the payout, used by the subsequent node.
         :param amount: The amount to put in the transaction, multiplier of base_amount.
-        :param base_amount: The base amount for the payouts.
+        :param base_amount: The base amount for the payout.
         """
         self.logger.info("Sending payout of %d (base: %d) to %s (cid: %s)", amount, base_amount, peer, circuit_id)
 
-        block = TriblerBandwidthBlock.create(
-            b'tribler_bandwidth',
-            {b'up': 0, b'down': amount},
-            self.bandwidth_wallet.trustchain.persistence,
-            self.my_peer.public_key.key_to_bin(),
-            link_pk=peer.public_key.key_to_bin())
-        block.sign(self.my_peer.key)
-        self.bandwidth_wallet.trustchain.persistence.add_block(block)
-
-        payload = PayoutPayload.from_half_block(block, circuit_id, base_amount)
-        packet = self._ez_pack(self._prefix, 23, [payload], False)
+        tx = self.bandwidth_community.construct_signed_transaction(peer, amount)
+        self.bandwidth_community.database.BandwidthTransaction.insert(tx)
+        payload = BandwidthTransactionPayload.from_transaction(tx, circuit_id, base_amount)
+        packet = self._ez_pack(self._prefix, 30, [payload], False)
         self.send_packet(peer, packet)
 
+    def on_payout(self, source_address: AddressType, data: bytes) -> None:
```

Comment:
Does it have to be here, or is it better to move it to `BandwidthAccountingCommunity`? I see almost no usages of Tunnels-related properties and methods here.

Reply:
Yeah this is not very pretty indeed, but I have good reasons to put it here. The main motivation is that a bandwidth payout in the tunnel community is slightly different from a direct payout (e.g., as the result of a libtorrent alert). Specifically, we want to include the circuit identifier when paying out to the next hop in a circuit so the subsequent hop knows which circuit the payment is associated with. Thus, we are using different IPv8 messages for these two forms of payouts.

I tried to think about a design where I just implemented `on_payout` once in `BandwidthAccountingCommunity` but that would result in code that was harder to read and maintain due to subtile differences between these two forms of payouts.